pallene/translator.lua

```diff
@@ -0,0 +1,62 @@
+local util = require "pallene.util"
+
+local translator = {}
+
+local Translator = util.Class()
+
+function Translator:init()
+    self.last_index = 1
+    self.partials = {}
+    return self
+end
+
+function Translator:add_previous(input, stop_index)
+    local partial = input:sub(self.last_index, stop_index)
+    table.insert(self.partials, partial)
+    self.last_index = stop_index + 1
+end
+
+-- TODO: Added test case for newlines, space, comments, and tabs.
+
+function Translator:add_whitespace(input, start_index, stop_index)
+    self:add_previous(input, start_index - 1)
+
+    local p = start_index
+    local q = start_index
+    while q <= stop_index do
+        if input:sub(q, q) == "\n" then
+            local partial = string.rep(" ", q - p - 1)
+            table.insert(self.partials, partial)
+            table.insert(self.partials, "\n")
+            p = q + 2
+        end
+        q = q + 1
+    end
+    local final_partial = string.rep(" ", q - p)
+    table.insert(self.partials, final_partial)
```

Comment:
I still think we should be able to replace this whole block of code by a call to `string.sub` followed by a call to `string.gsub`.

Reply:
Don't `sub` and `gsub` and create new strings every time we call them? Won't that be inefficient? I adopted this technique based on what is suggested in the Lua book.

Reply:
String.sub creates new strings but I think it should be OK because this is only for the type annotations, which are relatively small.

This is much less string creation and concatenation than we have in the coder.lua, which is a big collection of table.concat and util.render (which calls gsub behind the scenes). I suspect that we might even be doing some O(N^2) stuff over there. 😅